### PR TITLE
Change the context of wp_get_attachment_image_context

### PR DESCRIPTION
**Pull request by spacedmonkey**
Change context of wp_get_attachment_image_context in gallery shortcode to fix lazy loading and fetch priority. 

Trac ticket: https://core.trac.wordpress.org/ticket/58681

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by kt-12** [comment]
@spacedmonkey I liked the idea of modifying context of gallery shortcode. However, I was wondering if there is any chance of breaking or not executing any other logic that depends on `the_post_thumbnail` context either within the core or outside of the core. What do you feel?

**Comment by spacedmonkey** [comment]
> @spacedmonkey I liked the idea of modifying context of gallery shortcode. However, I was wondering if there is any chance of breaking or not executing any other logic that depends on `the_post_thumbnail` context either within the core or outside of the core. What do you feel?

I am not sure what you mean here? The context is change to `gallery_shortcode`  and the back to the default with the removal of the filter. For images that are the featured image then it would set to `the_post_thumbnail`. I don't see how this would break anything. 

As this is your ticket, I am going to leave this PR with you, see if you think it is a good part forward and maybe add some unit tests.

**Comment by joemcgill** [comment]
While this fixes the specific issue of the old gallery short code not being processed correctly, the bigger issue that this has made visible is that any images that are dynamically added to post content through a shortcode will also exhibit this bug. Rather than creating new context for any of these situations, I think we should take a look into why the rendered content isn't being processed correctly as content images. This is potentially a problem with the order in which shortcodes are processed compared with block content, but would need some investiagion.

**Comment by felixarntz** [comment]
Looking into this as well, I agree with @joemcgill's feedback. I believe [the fetchpriority refactor](https://core.trac.wordpress.org/changeset/56037) broke something that https://core.trac.wordpress.org/changeset/55825 initially fixed. Currently testing to verify whether that is indeed the case.

**Comment by felixarntz** [comment]
@spacedmonkey @joemcgill I left a comment outlining the issues I found in https://core.trac.wordpress.org/ticket/58681#comment:13

TL;DR even before this behavior changed, it was incorrect. The root problem is that shortcodes are only parsed after `wp_filter_content_tags()` has run. We could use approaches like this PR of #4799 to restore the previous behavior, but that's not really better than the way it behaves now.

**Comment by spacedmonkey** [comment]
Closing in favour of #4799
